src/tools/illink/src/linker/Linker/CustomSymbolWriter.cs

```diff
@@ -0,0 +1,117 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Text;
+
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace Mono.Linker
+{
+	internal sealed class CustomSymbolWriterProvider : ISymbolWriterProvider
+	{
+		readonly DefaultSymbolWriterProvider _defaultProvider = new DefaultSymbolWriterProvider ();
+		readonly bool _preserveSymbolPaths;
+
+		public CustomSymbolWriterProvider (bool preserveSymbolPaths) => this._preserveSymbolPaths = preserveSymbolPaths;
+
+		public ISymbolWriter GetSymbolWriter (ModuleDefinition module, string fileName)
+			=> new CustomSymbolWriter (_defaultProvider.GetSymbolWriter (module, fileName), module, _preserveSymbolPaths);
+
+		public ISymbolWriter GetSymbolWriter (ModuleDefinition module, Stream symbolStream)
+			=> new CustomSymbolWriter (_defaultProvider.GetSymbolWriter (module, symbolStream), module, _preserveSymbolPaths);
+	}
+
+	internal sealed class CustomSymbolWriter : ISymbolWriter
+	{
+		readonly ISymbolWriter _symbolWriter;
+		readonly ModuleDefinition _module;
+		readonly bool _preserveSymbolPaths;
+
+		internal CustomSymbolWriter (ISymbolWriter defaultWriter, ModuleDefinition module, bool preserveSymbolPaths)
+		{
+			_symbolWriter = defaultWriter;
+			_module = module;
+			_preserveSymbolPaths = preserveSymbolPaths;
+		}
+
+		public ImageDebugHeader GetDebugHeader ()
+		{
+			var header = _symbolWriter.GetDebugHeader ();
+			if (!_preserveSymbolPaths)
+				return header;
+
+			if (!header.HasEntries)
+				return header;
+
+			for (int i = 0; i < header.Entries.Length; i++) {
+				header.Entries [i] = ProcessEntry (header.Entries [i]);
+			}
+
+			return header;
+		}
+
+		ImageDebugHeaderEntry ProcessEntry (ImageDebugHeaderEntry entry)
+		{
+			if (entry.Directory.Type != ImageDebugType.CodeView)
+				return entry;
+
+			var reader = new BinaryReader (new MemoryStream (entry.Data));
+			var newDataStream = new MemoryStream ();
+			var writer = new BinaryWriter (newDataStream);
+
+			var sig = reader.ReadUInt32 ();
+			if (sig != 0x53445352)
+				return entry;
+
+			writer.Write (sig);
+			writer.Write (reader.ReadBytes (16)); // MVID
+			writer.Write (reader.ReadUInt32 ()); // Age
+
+			writer.Write (Encoding.UTF8.GetBytes (GetOriginalPdbPath ()));
+			writer.Write ((byte) 0);
+
+			var newData = newDataStream.ToArray ();
+
+			var directory = entry.Directory;
+			directory.SizeOfData = newData.Length;
+
+			return new ImageDebugHeaderEntry (directory, newData);
+		}
+
+		string GetOriginalPdbPath ()
+		{
+			if (!_module.HasDebugHeader)
+				return string.Empty;
+
+			var debugHeader = _module.GetDebugHeader ();
+			foreach (var entry in debugHeader.Entries) {
+				if (entry.Directory.Type != ImageDebugType.CodeView)
+					continue;
+
+				var reader = new BinaryReader (new MemoryStream (entry.Data));
+				var sig = reader.ReadUInt32 ();
+				if (sig != 0x53445352)
```

Comment:
nit: Could we make `0x53445352` a const with a descriptive name or put a comment for what this check indicates?